### PR TITLE
Preserve query string on redirect

### PR DESCRIPTION
**Pull request by c-w**
## Description

Mirror of https://github.com/evansd/whitenoise/pull/609

## Checklist

-   [x] Tests have been developed for bug fixes or new functionality.
-   [x] The changelog has been updated, if necessary.
-   [x] Documentation has been updated, if necessary.
-   [x] GitHub Issues closed by this PR have been linked.

<sub>By submitting this pull request I agree that all contributions comply with this project's open source license(s).</sub>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ Using the following categories, list your changes in this order:
 ### Added
 
 -   Support Python 3.13.
+-   Django `settings.py:SERVESTATIC_PRESERVE_QUERY_STRING_ON_REDIRECT` will forward the request query string when issuing redirects.
```
Comment (Archmonger) [inline CHANGELOG.md]: Given that there is no detriment to preserving query strings, I don't believe we should make a setting out of this.

It feels like the behavior of preserving query strings should always be on.

In legacy applications that didn't expect this behavior, the query strings just become extraneous data.

File: src/servestatic/responders.py [hunk 2]
```diff
@@ -315,12 +315,20 @@ def get_path_and_headers(self, request_headers):
...
         headers = list(headers.items()) if headers else []
         headers.append(("Location", quote(location.encode("utf8"))))
         self.response = Response(HTTPStatus.FOUND, headers, None)
+        self.preserve_query_string = preserve_query_string
 
     def get_response(self, method, request_headers):
+        query_string = request_headers.get("QUERY_STRING")
+        if query_string and self.preserve_query_string:
+            headers = list(self.response.headers)
+            name, value = headers[-1]
```
Comment (Archmonger) [inline src/servestatic/responders.py]: This line needs a comment for clarification - Specifically why is the query string always at position `headers[-1]`, and is this also true when using `ServeStatic` with ASGI?

File: tests/test_servestatic.py [hunk 3]
```diff
@@ -16,7 +16,7 @@
 import pytest
```
Comment (Archmonger) [inline tests/test_servestatic.py:16]: Needs an additional redirect test within `test_asgi.py`

File: CHANGELOG.md (hunk 1 above)
Comment (c-w) [reply CHANGELOG.md]: Removed the feature flag in fdfbf1d

File: src/servestatic/responders.py (hunk 2 above)
Comment (c-w) [reply src/servestatic/responders.py]: Made the header lookup more generic in 0db086d

File: tests/test_servestatic.py (hunk 3 above)
Comment (c-w) [reply tests/test_servestatic.py:16]: Added tests and fixed asgi behavior in fc0db6d

File: src/servestatic/asgi.py [hunk 4]
```diff
@@ -47,6 +47,7 @@ async def __call__(self, scope, receive, send):
         wsgi_headers = {
             "HTTP_" + key.decode().upper().replace("-", "_"): value.decode() for key, value in scope["headers"]
         }
+        wsgi_headers["QUERY_STRING"] = scope["query_string"].decode()
```
Comment (Archmonger) [inline src/servestatic/asgi.py:50]: Add a comment here and a better method would be to remove the duplicate value via `pop`

```suggestion
        # The query string header should not begin with `HTTP_`
        wsgi_headers["QUERY_STRING"] = wsgi_headers.pop("HTTP_QUERY_STRING")
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -14,6 +14,7 @@ Using the following categories, list your changes in this order:
 ### Added
 
 -   Support Python 3.13.
+-   Query strings are now preserved and forwarded to the location when redirecting.
```
Comment (Archmonger) [inline CHANGELOG.md]:
```suggestion
-   Query strings are now preserved during HTTP redirections.
```

File: src/servestatic/base.py [hunk 6]
```diff
@@ -236,4 +236,7 @@ def redirect(self, from_url, to_url):
             msg = f"Cannot handle redirect: {from_url} > {to_url}"
             raise ValueError(msg)
         headers = {"Cache-Control": f"max-age={self.max_age}, public"} if self.max_age is not None else {}
-        return Redirect(relative_url, headers=headers)
+        return Redirect(
+            relative_url,
+            headers=headers,
+        )
```
Comment (Archmonger) [inline src/servestatic/base.py]: Revert unneeded formatting change

```suggestion
        return Redirect(relative_url, headers=headers)
```

File: tests/test_asgi.py [hunk 7]
```diff
@@ -47,6 +53,14 @@ def test_get_js_static_file(application, test_files):
     assert send.headers[b"content-length"] == str(len(test_files.js_content)).encode()
 
 
+def test_redirect_preserves_query_string(application, test_files):
+    scope = AsgiScopeEmulator({"path": "/static/with-index", "query_string": b"v=1"})
+    receive = AsgiReceiveEmulator()
+    send = AsgiSendEmulator()
+    asyncio.run(application(scope, receive, send))
+    assert send.headers[b"location"] == b"with-index/?v=1"
```
Comment (Archmonger) [inline tests/test_asgi.py]: Make this test slightly more robust by adding an second parameter into the query string, such as `with-index/?v=1&x=2`

```suggestion
    scope = AsgiScopeEmulator({"path": "/static/with-index", "query_string": b"v=1&x=2"})
    receive = AsgiReceiveEmulator()
    send = AsgiSendEmulator()
    asyncio.run(application(scope, receive, send))
    assert send.headers[b"location"] == b"with-index/?v=1&x=2"
```

**Review by Archmonger — CHANGES_REQUESTED** [review]

File: src/servestatic/asgi.py (hunk 4 above)
Comment (c-w) [reply src/servestatic/asgi.py:50]: Could you be more specific? From what I can tell, query string isn't a header as it's set on the scope directly under the key `query_string` (see [asgi spec](https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope)).

Comment (Archmonger) [reply src/servestatic/asgi.py:50]: Ah my bad, you're correct.

File: CHANGELOG.md [hunk 8]
```diff
@@ -14,6 +14,7 @@ Using the following categories, list your changes in this order:
 ### Added
 
 -   Support Python 3.13.
+-   Query strings are now preserved during HTTP redirections.
```
Comment (Archmonger) [inline CHANGELOG.md]: Small fix to quiet the spell checker CI

```suggestion
-   Query strings are now preserved during HTTP redirection.
```

**Review by Archmonger — APPROVED** [review]

**Comment by Archmonger** [comment]
Looks good, will create a new release tomorrow.

**Pull request merged**